File: frontend/src/api/knowledge.ts
import { apiClient } from "./client";


export type KnowledgeDocRead = {
  id: number;
  title: string;
  file_name: string;
  file_type: string;
  file_path: string;
  content: string;
  doc_type: string;
  status: string;
  uploaded_by: number;
  error_message: string | null;
  chunks_count: number;
  created_at: string;
  updated_at: string;
};

export type KnowledgeChunkRead = {
  id: number;
  doc_id: number;
  chunk_index: number;
  content: string;
  metadata_json: Record<string, unknown>;
  embedding_id: string | null;
  created_at: string;
  updated_at: string;
};

export type KnowledgeSearchPayload = {
  query: string;
  top_k?: number;
};

export type KnowledgeSearchResult = {
  doc_id: number;
  chunk_id: number;
  chunk_index: number;
  content_preview: string;
  score: number;
  embedding_id: string | null;
};

export async function listKnowledgeDocs() {
  const response = await apiClient.get<KnowledgeDocRead[]>("/knowledge/docs");
  return response.data;
}

export async function getKnowledgeDoc(docId: number) {
  const response = await apiClient.get<KnowledgeDocRead>(`/knowledge/docs/${docId}`);
  return response.data;
}

export async function listKnowledgeChunks(docId: number) {
  const response = await apiClient.get<KnowledgeChunkRead[]>(`/knowledge/docs/${docId}/chunks`);
  return response.data;
}

export async function uploadKnowledgeDoc(payload: { title: string; file: File }) {
  const formData = new FormData();
  formData.append("title", payload.title);
  formData.append("file", payload.file);

  const response = await apiClient.post<KnowledgeDocRead>("/knowledge/upload", formData);
  return response.data;
}

export async function searchKnowledge(payload: KnowledgeSearchPayload) {
  const response = await apiClient.post<KnowledgeSearchResult[]>("/knowledge/search", {
    query: payload.query,
    top_k: payload.top_k ?? 5,
  });
  return response.data;
}

File: backend/app/schemas/knowledge.py
from pydantic import BaseModel, ConfigDict, Field

from app.utils.datetime_utils import UTCDatetime


class KnowledgeChunkRead(BaseModel):
    id: int
    doc_id: int
    chunk_index: int
    content: str
    metadata_json: dict
    embedding_id: str | None
    created_at: UTCDatetime
    updated_at: UTCDatetime

    model_config = ConfigDict(from_attributes=True)


class KnowledgeDocRead(BaseModel):
    id: int
    title: str
    file_name: str
    file_type: str
    file_path: str
    content: str
    doc_type: str
    status: str
    uploaded_by: int
    error_message: str | None
    chunks_count: int = 0
    created_at: UTCDatetime
    updated_at: UTCDatetime

    model_config = ConfigDict(from_attributes=True)


class KnowledgeSearchRequest(BaseModel):
    query: str = Field(min_length=1, max_length=2000)
    top_k: int = Field(default=5, ge=1, le=20)


class KnowledgeSearchResult(BaseModel):
    doc_id: int
    chunk_id: int
    chunk_index: int
    content_preview: str
    score: float
    embedding_id: str | None

    model_config = ConfigDict(from_attributes=True)
